app/controllers/time_slot_controller.rb

```diff
@@ -54,39 +53,24 @@ def cancel
 
   def index
     @appointment = current_patient.current_appointment
-    @ubs = @appointment.try(:ubs)
-
+    @ubs = @appointment&.ubs
     @patient_can_schedule = current_patient.can_schedule?
-
     return render_vaccinated if current_patient.vaccinated?
 
-    @gap_in_days = slot_params[:gap_in_days].to_i || 0
-    @current_day = Time.zone.now + @gap_in_days.days
-
-    @time_slots = {}
+    @gap_in_days = gap_in_days
+    @current_day = [Time.zone.now.at_beginning_of_day + @gap_in_days.days, Time.zone.now].max # prevent users from scheduling in the past
 
     last_appointment = current_patient.last_appointment
     return if last_appointment&.second_dose? && @current_day.to_date < last_appointment.start.to_date
 
-    Ubs.where(active: true).each do |ubs|
-      slots = []
-
-      if @gap_in_days <= TimeSlotController::SLOTS_WINDOW_IN_DAYS && @gap_in_days >= 0
-        if @current_day.today?
-          appointments = Appointment.where(start: @current_day..@current_day.end_of_day, ubs: ubs, patient_id: nil)
-        else
-          appointments = Appointment.where(start: @current_day.at_beginning_of_day..@current_day.end_of_day, ubs: ubs, patient_id: nil)
-        end
-
-        next unless appointments.exists?
-
-        appointments.order(:start).each do |appointment|
-          slots << { slot_start: appointment.start, slot_end: appointment.end }
```

Comment:
I see we are now sending to the front the whole appointment data as a ruby object (we were already doing that for the UBS). Don't think it's a problem, but just a double-check, is there any scenario where we would be using too much memory? WDYT @edrd-f ?

Reply:
Yes, it's returned as a hollowed out Ruby object as there are only `ubs_id`, `start` and `end`. Due to the way Ruby behaves on 2.5+ it shouldn't matter too. Its way more performant then it is in its current state with N+1 queries to the DB.

Reply:
There shouldn't be a big increase in memory because the query is selecting only `:ubs_id, :start, :end`.